src/main/ts/core/Changelog.ts

```diff
@@ -0,0 +1,361 @@
+}
+
+interface ReleaseMeta {
+  version: Version;
+  date: DateTime;
+}
+
+interface Release extends ChangelogFragment, Offset {
+  header: Offset;
+  meta?: ReleaseMeta;
+}
+
+export interface Changelog {
+  source: string;
+  releases: Release[];
+  preamble: Offset;
+  links: Offset;
+}
+
+interface TextLines {
+  text: string;
+  lines: number[];
+}
+
+interface Heading {
+  header: commonmark.Node;
+  preamble: commonmark.Node[];
+  subheadings: Heading[];
+}
+
+interface Top {
+  header?: undefined;
+  preamble: commonmark.Node[];
+  subheadings: Heading[];
+}
+
+const findLineStarts = (text: string): TextLines => {
+  const lines = [ 0 ];
+  let seenCR = false;
+  let seenLF = false;
+  for (let i = 0; i < text.length; i++) {
+    const c = text.charAt(i);
+    if (c === '\n') {
+      if (seenLF) {
+        // add line start and reset
+        lines.push(i);
+        seenCR = false;
+        seenLF = false;
+      }
+      seenLF = true;
+    } else if (c === '\r') {
+      if (seenCR) {
+        // add line start and reset
+        lines.push(i);
+        seenCR = false;
+        seenLF = false;
+      }
+      seenCR = true;
+    } else {
+      if (seenLF || seenCR) {
+        // add line start and reset
+        lines.push(i);
+        seenLF = false;
+        seenCR = false;
+      }
+    }
+  }
+  return { text, lines };
+};
+
+const posToOffset = (lines: number[], position: [number, number]) => {
+  // both line and column count from 1
+  const [ line, column ] = position;
+  return lines[line - 1] + (column - 1);
+};
+
+// the range including just the columns of the node
+const columnRange = (lines: number[], node: commonmark.Node): Offset => {
+  // this is the inclusive range
+  const [ startPos, endPos ] = node.sourcepos;
+  const start = posToOffset(lines, startPos);
+  const end = posToOffset(lines, endPos);
+  return { start, end };
+};
+
+// the range including the whole lines that the node is in
+const blockRange = (source: TextLines, node: commonmark.Node): Offset => {
+  const [[ startLine ], [ endLine ]] = node.sourcepos;
+  const start = posToOffset(source.lines, [ startLine, 1 ]);
+  const end = (endLine < source.lines.length ? posToOffset(source.lines, [ endLine + 1, 1 ]) : source.text.length) - 1;
+  return { start, end };
+};
+
+const extractText = (source: TextLines, node: commonmark.Node): string => {
+  const range = columnRange(source.lines, node);
+  return source.text.substring(range.start, range.end + 1);
+};
+
+const pos = (node: commonmark.Node): string => {
+  if (node.sourcepos !== undefined) {
+    const [[ startLine, startColumn ]] = node.sourcepos;
+    return ` (line: ${startLine} column: ${startColumn})`;
+  } else {
+    return '';
+  }
+};
+
+const parseHeadings = (changelog: string): Top => {
+  const data: Top = { preamble: [], subheadings: [] };
+  const reader = new commonmark.Parser();
+  const doc = reader.parse(changelog);
+  const stack: (Heading | Top)[] = [ data ];
```

Comment:
Not sure about this particularly, but can you add a comment describing the invariants / reasoning behind this stack and the loop below?
```suggestion
  // Always of the form [ top, h1, h2, h3, ... ] so that we can tell which is the latest heading (of any level) to attach new nodes to
  const stack: (Heading | Top)[] = [ data ];
```

Reply:
Also maybe call it `headingStack` or something? idk

Reply:
Sure.

Reply:
Done